buildSrc/src/main/java/org/elasticsearch/gradle/DockerUtils.java

```diff
@@ -0,0 +1,228 @@
+        private String version = null;
+        private Result lastCommand = null;
+
+        /**
+         * Indicates whether Docker is available and meets the required criteria.
+         * @return true if, and only if, Docker is:
+         * <ul>
+         *     <li>Installed</li>
+         *     <li>Executable</li>
+         *     <li>Is at least version 17.05</li>
+         *     <li>Can execute a command that requires privileges</li>
+         * </ul>
+         */
+        public boolean isAvailable() {
+            return isAvailable;
+        }
+
+        /**
+         * @return true if the installed Docker version is &gt;= 17.05
+         */
+        public boolean isVersionHighEnough() {
+            return isVersionHighEnough;
+        }
+
+        /**
+         * @return the path to the Docker CLI, or null
+         */
+        public String getPath() {
+            return path;
+        }
+
+        /**
+         * @return The installed Docker version, or null
+         */
+        public String getVersion() {
+            return version;
+        }
+
+        /**
+         * @return Information about the last command executes while probing Docker, or null.
+         */
+        public Result getLastCommand() {
+            return lastCommand;
+        }
+    }
+
+    // package-private for testing
+    /**
+     * Given the string that <code>docker --version</code> prints, extract the docker version from it.
+     * @param dockerVersion the version string to parse
+     * @return the Docker version
+     */
+    static String extractVersion(String dockerVersion) {
+        final Pattern dockerVersionPattern = Pattern.compile("Docker version (\\d+\\.\\d+\\.\\d+(?:-[a-zA-Z0-9]+)?), build [0-9a-f]{7,40}");
+
+        final Matcher matcher = dockerVersionPattern.matcher(dockerVersion);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Can't extract version from: [" + dockerVersion + "]");
+        }
+
+        return matcher.group(1);
+    }
+
+    // package-private for testing
+    /**
+     * Checks whether the supplied Docker version is &gt;= 17.05
+     * @param dockerVersion the version to check
+     * @return true if the version is high enough
+     */
+    static boolean checkVersion(String dockerVersion) {
+        final String[] majorMinor = dockerVersion.split("\\.");
+
+        return (Integer.parseInt(majorMinor[0]) > 17
+            || (Integer.parseInt(majorMinor[0]) == 17 && Integer.parseInt(majorMinor[1]) >= 5));
+    }
+
+    /**
+     * Runs a command and captures the exit code, standard output and standard error.
+     * @param args the command and any arguments to execute
+     * @return a object that captures the result of running the command. If an exception occurring
+     * while running the command, or the process was killed after reaching the 10s timeout,
+     * then the exit code will be -1.
+     */
+    private static Result runCommand(String... args) {
```

Comment:
This command running stuff seems overly complex to support generic use, yet it's embedded here as an inner-class. I feel like we should either a) just make this a lot simpler, just to solve the "what docker version am I using use-case" or b) rip it out so it can be used elsewhere.